sdk/javascript/wasm/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+			hash_mode: hash_mode,
+			_keccak: Keccak512::new(),
+			_sha2: Sha512::new()
+		}
+	}
+
+	fn update<T: AsRef<[u8]> + ?Sized>(&mut self, input: &T) {
+		if HashMode::Keccak == self.hash_mode {
+			self._keccak.update(&input);
+		} else {
+			self._sha2.update(&input);
+		}
+	}
+
+	fn finalize_into(self, out: &mut [u8; HASH_LENGTH]) {
+		if HashMode::Keccak == self.hash_mode {
+			out.copy_from_slice(&self._keccak.finalize().as_slice());
+		} else {
+			out.copy_from_slice(&self._sha2.finalize().as_slice());
+		}
+	}
+}
+
+// endregion
+
+// region helpers
+
+fn hash_private_key(hash_mode: HashMode, sk: &[u8; SECRET_KEY_LENGTH]) -> [u8; HASH_LENGTH] {
+	let mut hasher = Hasher512::new(hash_mode);
+	hasher.update(&sk);
+
+	let mut hash: [u8; HASH_LENGTH] = [0u8; HASH_LENGTH];
+	hasher.finalize_into(&mut hash);
+	hash
+}
+
+fn copy_and_clamp_private_hash(private_hash: &[u8; HASH_LENGTH]) -> Scalar {
+	let mut bits: [u8; SECRET_KEY_LENGTH] = [0; SECRET_KEY_LENGTH];
+	bits.copy_from_slice(&private_hash[..SECRET_KEY_LENGTH]);
+
+	bits[0] &= 248;
+	bits[31] &= 127;
+	bits[31] |= 64;
+
+	Scalar::from_bits(*&mut bits)
+}
+
+fn hash_to_scalar(builder: Hasher512) -> Scalar {
+	let mut hash = [0u8; HASH_LENGTH];
+	builder.finalize_into(&mut hash);
+	Scalar::from_bytes_mod_order_wide(&hash)
+}
+
+fn generate_nonce(hash_mode: HashMode, private_hash: &[u8; HASH_LENGTH], message: &[u8]) -> Scalar {
+	// hash half of priv key hash
+	let mut builder = Hasher512::new(hash_mode);
+	builder.update(&private_hash[32..64]);
+	builder.update(&message);
+
+	hash_to_scalar(builder)
+}
+
+// endregion
+
+// region implementations
+
+/// Generates public key given secret key.
+pub fn crypto_sign_keypair_unboxed(hash_mode: HashMode, sk: &[u8; SECRET_KEY_LENGTH]) -> [u8; PUBLIC_KEY_LENGTH] {
+	set_panic_hook();
+
+	let private_hash = hash_private_key(hash_mode, sk);
+	let key = copy_and_clamp_private_hash(&private_hash);
+	let point = &key * &curve25519_dalek::constants::ED25519_BASEPOINT_TABLE;
+	let compressed = point.compress();
+	compressed.to_bytes()
+}
+
+/// Signs message.
+pub fn crypto_private_sign_unboxed(hash_mode: HashMode, sk: &[u8; SECRET_KEY_LENGTH], message: &[u8]) -> [u8; SIGNATURE_LENGTH] {
+	set_panic_hook();
+
+	let private_hash = hash_private_key(hash_mode, sk);
+
+	// R = rModQ * base point
+	let r = generate_nonce(hash_mode, &private_hash, &message);
+	#[allow(non_snake_case)]
+	let R: CompressedEdwardsY = (&r * &curve25519_dalek::constants::ED25519_BASEPOINT_TABLE).compress();
+
+	// h = H(encodedR || public || data)
+	let public_key = crypto_sign_keypair_unboxed(hash_mode, &sk);
+
+	let mut builder_h = Hasher512::new(hash_mode);
+	builder_h.update(R.as_bytes());
+	builder_h.update(&public_key);
+	builder_h.update(&message);
+	let h = hash_to_scalar(builder_h);
+
+	// S = (r + h * a) mod group order
+	let a = copy_and_clamp_private_hash(&private_hash);
+	#[allow(non_snake_case)]
+	let S = &(&h * &a) + &r;
+
+	// serialize and concat R S
```

Comment:
we decided not to do additional S check, right?

Reply:
which additional S check?